gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaDeserializerExtractor.java

```diff
@@ -0,0 +1,193 @@
+
+import kafka.message.MessageAndOffset;
+import kafka.utils.VerifiableProperties;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+import gobblin.configuration.WorkUnitState;
+import gobblin.metrics.kafka.KafkaSchemaRegistry;
+import gobblin.metrics.kafka.SchemaRegistryException;
+import gobblin.util.PropertiesUtils;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+
+/**
+ * <p>
+ *   Extension of {@link KafkaExtractor} that wraps Kafka's {@link Deserializer} API. Kafka's {@link Deserializer} provides
+ *   a generic way of converting Kafka {@link kafka.message.Message}s to {@link Object}. Typically, a {@link Deserializer}
+ *   will be used along with a {@link org.apache.kafka.common.serialization.Serializer} which is responsible for converting
+ *   an {@link Object} to a Kafka {@link kafka.message.Message}. These APIs are useful for reading and writing to Kafka,
+ *   since Kafka is primarily a byte oriented system.
+ * </p>
+ *
+ * <p>
+ *   This class wraps the {@link Deserializer} API allowing any existing classes that implement the {@link Deserializer}
+ *   API to integrate with seamlessly with Gobblin. The deserializer can be specified in the following ways:
+ *
+ *   <ul>
+ *     <li>{@link #KAFKA_DESERIALIZER_TYPE} can be used to specify a pre-defined enum from {@link Deserializers}</li>
+ *     <li>{@link #KAFKA_DESERIALIZER_CLASS} can be used to specify the full-qualified name of a {@link Class} that
+ *     defines the {@link Deserializer} interface. If this property is used a {@link KafkaSchemaRegistry} must also
+ *     be specified using the {@link KafkaSchemaRegistry#KAFKA_SCHEMA_REGISTRY_CLASS} config key</li>
+ *   </ul>
+ * </p>
+ */
+@Getter(AccessLevel.PACKAGE)
+public class KafkaDeserializerExtractor extends KafkaExtractor<Object, Object> {
+
+  public static final String KAFKA_DESERIALIZER_TYPE = "kafka.deserializer.type";
+  public static final String KAFKA_DESERIALIZER_CLASS = "kafka.deserializer.class";
+
+  private static final String CONFLUENT_SCHEMA_REGISTRY_URL = "schema.registry.url";
+
+  private final Deserializer<?> kafkaDeserializer;
+  private final KafkaSchemaRegistry<?, ?> kafkaSchemaRegistry;
+
+  public KafkaDeserializerExtractor(WorkUnitState state) throws ReflectiveOperationException {
+    this(state, getDeserializer(getProps(state)),
+        getKafkaSchemaRegistry(getProps(state)));
+  }
+
+  @VisibleForTesting
+  KafkaDeserializerExtractor(WorkUnitState state, Deserializer<?> kafkaDeserializer,
+      KafkaSchemaRegistry<?, ?> kafkaSchemaRegistry) {
+    super(state);
+    this.kafkaDeserializer = kafkaDeserializer;
+    this.kafkaSchemaRegistry = kafkaSchemaRegistry;
+  }
+
+  @Override
+  protected Object decodeRecord(MessageAndOffset messageAndOffset) throws IOException {
+    return this.kafkaDeserializer.deserialize(this.topicName, getBytes(messageAndOffset.message().payload()));
```

Comment:
Since the `Deserializer` interface doesn't allow passing in a schema registry object, it is up to the `Deserializer` implementation to create a schema registry using some configuration properties.

So in the case of the Confluent `KafkaAvroDeserializer` it will construct and use its own schema registry. Separate from the one used in the `KafkaDeserializerExtractor`. So two different schema registries will be created, which is unfortunately unavoidable in this scenario.